# EDA (Exploratory Data Analysis)


In [7]:
!pip install sklearn
!pip install pandas
!pip install numpy

  Created wheel for sklearn: filename=sklearn-0.0-py2.py3-none-any.whl size=1316 sha256=bc15b2bc44d7416a7d6de91dce9b5fc85dd7fd9771935c4f09ec245b59f97604
  Stored in directory: c:\users\ryang\appdata\local\pip\cache\wheels\22\0b\40\fd3f795caaa1fb4c6cb738bc1f56100be1e57da95849bfc897
Successfully built sklearn


In [9]:
!pip install plotly

In [10]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px
from sklearn import datasets

%matplotlib inline 
%matplotlib notebook 

In [134]:
from pandas import DataFrame
from pandas import Series

## 데이터 살펴보기

In [87]:
data = pd.read_csv(
    "C:\\Users\\ryang\\Downloads\\파이널프로젝트_RAW_210329_210926.csv")

In [88]:
data

id  customer_id  course_id     type      state  \
0        970634     504760.0   201435.0    ORDER  CANCELLED   
1        970650     432862.0   203178.0    ORDER  CANCELLED   
2        970657      72297.0   204246.0  PAYMENT  COMPLETED   
3        970656      72297.0   204246.0    ORDER  CANCELLED   
4        970658     478028.0   201797.0  PAYMENT  COMPLETED   
...         ...          ...        ...      ...        ...   
105414  1364995     640697.0   205365.0  PAYMENT  CANCELLED   
105415  1364998     670896.0   204159.0  PAYMENT  CANCELLED   
105416  1364996     640697.0   205365.0  PAYMENT  COMPLETED   
105417  1365005     670896.0   204159.0  PAYMENT  CANCELLED   
105418  1365007     610626.0   206720.0  PAYMENT  COMPLETED   

                           course_title category_title  format  \
0             올인원 패키지 : 알고리즘 기술면접 완전 정복          프로그래밍     올인원   
1           초격차 패키지 : 한번에 끝내는 영상 촬영과 편집          영상/3D     올인원   
2                  시크릿코드 : 프론트엔드 실무 완성편          프로그래밍  스쿨 온라인   
3                  시크릿코드 : 프론트엔드 실무 완성편          프로그래밍  스쿨 온라인   
4         올인원 패키지 : 직장인 필수 스킬 3종 세트 MAX         업무 생산성     올인원   
...                                 ...            ...     ...   
105414  초격차 패키지 : 한 번에 끝내는 유튜브 크리에이터 되기          영상/3D     올인원   
105415              부동산 디벨로퍼 아카데미 : 인허가         부동산/금융     올인원   
105416  초격차 패키지 : 한 번에 끝내는 유튜브 크리에이터 되기          영상/3D     올인원   
105417              부동산 디벨로퍼 아카데미 : 인허가         부동산/금융     올인원   
105418               용호수의 돈 버는 실전 영상 제작          영상/3D     올인원   

               completed_at  transaction_amount               coupon_title  \
0       2021-03-28 15:00:11               90500                        NaN   
1       2021-03-28 15:16:53               96000                        NaN   
2       2021-03-28 15:21:43              171810      [10% 할인] 시크릿코드 실무 완성편   
3       2021-03-28 15:21:43              171810                        NaN   
4       2021-03-28 15:21:54               97500   [웰컴쿠폰] 올인원 패키지 2만원 할인 쿠폰   
...                     ...                 ...                        ...   
105414  2021-09-25 14:10:42               96000  [WELCOME] 온라인 강의 2만원 할인쿠폰   
105415  2021-09-25 14:16:33              189000                        NaN   
105416  2021-09-25 14:17:41               96000  [WELCOME] 온라인 강의 2만원 할인쿠폰   
105417  2021-09-25 14:19:56              189000                        NaN   
105418  2021-09-25 14:25:47              139000  [WELCOME] 온라인 강의 3만원 할인쿠폰   

        coupon_discount_amount  sale_price  tax_free_amount      pg method  \
0                          NaN    110500.0         110500.0     NaN  VBANK   
1                          NaN    116000.0         116000.0     NaN   CARD   
2                      19090.0    190900.0         190900.0     NaN   CARD   
3                          NaN    190900.0         190900.0     NaN   CARD   
4                      20000.0    117500.0         117500.0     NaN   CARD   
...                        ...         ...              ...     ...    ...   
105414                 20000.0    116000.0         116000.0  INICIS  TRANS   
105415                     NaN    189000.0         189000.0  INICIS   CARD   
105416                 20000.0    116000.0         116000.0  INICIS  VBANK   
105417                     NaN    189000.0         189000.0  INICIS   CARD   
105418                 30000.0    169000.0         169000.0   NAVER  POINT   

       subcategory_title  marketing_start_at  
0                개발자 커리어                 NaN  
1                  영상 편집  2020-12-02 0:00:00  
2                    NaN  2021-03-04 0:00:00  
3                    NaN  2021-03-04 0:00:00  
4                PPT/보고서  2019-11-14 0:00:00  
...                  ...                 ...  
105414             영상 편집  2021-06-15 0:00:00  
105415            부동산 개발  2021-02-10 0:00:00  
105416             영상 편집  2021-06-15 0:00:00  
105417            부동산 개발  2021-02-10 0:00:00  
105418             영상 편집  2021-08-17 0:00:00  

[105419 rows x 18 column

In [89]:
data.describe()

id    customer_id      course_id  transaction_amount  \
count  1.054190e+05  105416.000000  100212.000000        1.054190e+05   
mean   1.147054e+06  417248.252742  203868.586327        1.464814e+05   
std    1.125898e+05  206414.496648    1534.583331        7.195061e+05   
min    5.679930e+05      16.000000    2516.000000       -4.144074e+06   
25%    1.048940e+06  256401.000000  203100.000000        8.350000e+04   
50%    1.135186e+06  501046.000000  203787.000000        1.175000e+05   
75%    1.240472e+06  566595.000000  204775.000000        1.590000e+05   
max    1.381657e+06  672308.000000  207742.000000        1.134000e+08   

       coupon_discount_amount    sale_price  tax_free_amount  
count            3.181700e+04  1.042910e+05     1.042910e+05  
mean             4.972172e+04  1.610808e+05     1.575551e+05  
std              5.624467e+04  3.955298e+05     3.898133e+05  
min              0.000000e+00  1.000000e+00     0.000000e+00  
25%              2.000000e+04  1.090000e+05     1.090000e+05  
50%              2.100000e+04  1.320000e+05     1.320000e+05  
75%              6.915000e+04  1.710000e+05     1.710000e+05  
max              2.000000e+06  9.094000e+07     9.094000e+07

### 데이터 type 확인

In [90]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 105419 entries, 0 to 105418
Data columns (total 18 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   id                      105419 non-null  int64  
 1   customer_id             105416 non-null  float64
 2   course_id               100212 non-null  float64
 3   type                    105419 non-null  object 
 4   state                   105419 non-null  object 
 5   course_title            105403 non-null  object 
 6   category_title          105395 non-null  object 
 7   format                  105403 non-null  object 
 8   completed_at            105419 non-null  object 
 9   transaction_amount      105419 non-null  int64  
 10  coupon_title            31817 non-null   object 
 11  coupon_discount_amount  31817 non-null   float64
 12  sale_price              104291 non-null  float64
 13  tax_free_amount         104291 non-null  float64
 14  pg                  

### 결측값/결측비율 확인

In [91]:
print('───[결측값]────────────────────────')
print(data.isna().sum())
print()
print('───[결측비율]────────────────────────')
print(data.isna().sum() / len(data) *100)

───[결측값]────────────────────────
id                            0
customer_id                   3
course_id                  5207
type                          0
state                         0
course_title                 16
category_title               24
format                       16
completed_at                  0
transaction_amount            0
coupon_title              73602
coupon_discount_amount    73602
sale_price                 1128
tax_free_amount            1128
pg                        15075
method                    13096
subcategory_title         19829
marketing_start_at        17979
dtype: int64

───[결측비율]────────────────────────
id                         0.000000
customer_id                0.002846
course_id                  4.939337
type                       0.000000
state                      0.000000
course_title               0.015178
category_title             0.022766
format                     0.015178
completed_at               0.000000
transaction_amount 

### 불필요한 컬럼 제거

In [92]:
data.drop(['tax_free_amount'], axis =1)

id  customer_id  course_id     type      state  \
0        970634     504760.0   201435.0    ORDER  CANCELLED   
1        970650     432862.0   203178.0    ORDER  CANCELLED   
2        970657      72297.0   204246.0  PAYMENT  COMPLETED   
3        970656      72297.0   204246.0    ORDER  CANCELLED   
4        970658     478028.0   201797.0  PAYMENT  COMPLETED   
...         ...          ...        ...      ...        ...   
105414  1364995     640697.0   205365.0  PAYMENT  CANCELLED   
105415  1364998     670896.0   204159.0  PAYMENT  CANCELLED   
105416  1364996     640697.0   205365.0  PAYMENT  COMPLETED   
105417  1365005     670896.0   204159.0  PAYMENT  CANCELLED   
105418  1365007     610626.0   206720.0  PAYMENT  COMPLETED   

                           course_title category_title  format  \
0             올인원 패키지 : 알고리즘 기술면접 완전 정복          프로그래밍     올인원   
1           초격차 패키지 : 한번에 끝내는 영상 촬영과 편집          영상/3D     올인원   
2                  시크릿코드 : 프론트엔드 실무 완성편          프로그래밍  스쿨 온라인   
3                  시크릿코드 : 프론트엔드 실무 완성편          프로그래밍  스쿨 온라인   
4         올인원 패키지 : 직장인 필수 스킬 3종 세트 MAX         업무 생산성     올인원   
...                                 ...            ...     ...   
105414  초격차 패키지 : 한 번에 끝내는 유튜브 크리에이터 되기          영상/3D     올인원   
105415              부동산 디벨로퍼 아카데미 : 인허가         부동산/금융     올인원   
105416  초격차 패키지 : 한 번에 끝내는 유튜브 크리에이터 되기          영상/3D     올인원   
105417              부동산 디벨로퍼 아카데미 : 인허가         부동산/금융     올인원   
105418               용호수의 돈 버는 실전 영상 제작          영상/3D     올인원   

               completed_at  transaction_amount               coupon_title  \
0       2021-03-28 15:00:11               90500                        NaN   
1       2021-03-28 15:16:53               96000                        NaN   
2       2021-03-28 15:21:43              171810      [10% 할인] 시크릿코드 실무 완성편   
3       2021-03-28 15:21:43              171810                        NaN   
4       2021-03-28 15:21:54               97500   [웰컴쿠폰] 올인원 패키지 2만원 할인 쿠폰   
...                     ...                 ...                        ...   
105414  2021-09-25 14:10:42               96000  [WELCOME] 온라인 강의 2만원 할인쿠폰   
105415  2021-09-25 14:16:33              189000                        NaN   
105416  2021-09-25 14:17:41               96000  [WELCOME] 온라인 강의 2만원 할인쿠폰   
105417  2021-09-25 14:19:56              189000                        NaN   
105418  2021-09-25 14:25:47              139000  [WELCOME] 온라인 강의 3만원 할인쿠폰   

        coupon_discount_amount  sale_price      pg method subcategory_title  \
0                          NaN    110500.0     NaN  VBANK           개발자 커리어   
1                          NaN    116000.0     NaN   CARD             영상 편집   
2                      19090.0    190900.0     NaN   CARD               NaN   
3                          NaN    190900.0     NaN   CARD               NaN   
4                      20000.0    117500.0     NaN   CARD           PPT/보고서   
...                        ...         ...     ...    ...               ...   
105414                 20000.0    116000.0  INICIS  TRANS             영상 편집   
105415                     NaN    189000.0  INICIS   CARD            부동산 개발   
105416                 20000.0    116000.0  INICIS  VBANK             영상 편집   
105417                     NaN    189000.0  INICIS   CARD            부동산 개발   
105418                 30000.0    169000.0   NAVER  POINT             영상 편집   

        marketing_start_at  
0                      NaN  
1       2020-12-02 0:00:00  
2       2021-03-04 0:00:00  
3       2021-03-04 0:00:00  
4       2019-11-14 0:00:00  
...                    ...  
105414  2021-06-15 0:00:00  
105415  2021-02-10 0:00:00  
105416  2021-06-15 0:00:00  
105417  2021-02-10 0:00:00  
105418  2021-08-17 0:00:00  

[105419 rows x 17 columns]

### 실결제 완료 (state = COMPLETED) 정보만 가져옴

In [93]:
fc = data[data['state'] == 'COMPLETED']
fc

id  customer_id  course_id     type      state  \
2        970657      72297.0   204246.0  PAYMENT  COMPLETED   
4        970658     478028.0   201797.0  PAYMENT  COMPLETED   
8        970669     478033.0   203178.0  PAYMENT  COMPLETED   
10       970671     469860.0   203644.0  PAYMENT  COMPLETED   
14       970676     478035.0   203720.0  PAYMENT  COMPLETED   
...         ...          ...        ...      ...        ...   
105411  1364969     669238.0   204286.0  PAYMENT  COMPLETED   
105412  1364976     384003.0   203178.0  PAYMENT  COMPLETED   
105413  1364991     670891.0   203935.0  PAYMENT  COMPLETED   
105416  1364996     640697.0   205365.0  PAYMENT  COMPLETED   
105418  1365007     610626.0   206720.0  PAYMENT  COMPLETED   

                                          course_title category_title  format  \
2                                 시크릿코드 : 프론트엔드 실무 완성편          프로그래밍  스쿨 온라인   
4                        올인원 패키지 : 직장인 필수 스킬 3종 세트 MAX         업무 생산성     올인원   
8                          초격차 패키지 : 한번에 끝내는 영상 촬영과 편집          영상/3D     올인원   
10      올인원 패키지 : 권오상의 금융 아카데미 - 권오상의 재무제표 분석 기초(평생소장)         부동산/금융     올인원   
14                         초격차 패키지 : 한 번에 끝내는 프론트엔드 개발          프로그래밍     올인원   
...                                                ...            ...     ...   
105411                      초격차 패키지 : 한번에 끝내는 3D 디자인 툴            디자인     올인원   
105412                     초격차 패키지 : 한번에 끝내는 영상 촬영과 편집          영상/3D     올인원   
105413                     초격차 패키지 : 한 번에 끝내는 파이썬 웹 개발          프로그래밍     올인원   
105416                 초격차 패키지 : 한 번에 끝내는 유튜브 크리에이터 되기          영상/3D     올인원   
105418                              용호수의 돈 버는 실전 영상 제작          영상/3D     올인원   

               completed_at  transaction_amount               coupon_title  \
2       2021-03-28 15:21:43              171810      [10% 할인] 시크릿코드 실무 완성편   
4       2021-03-28 15:21:54               97500   [웰컴쿠폰] 올인원 패키지 2만원 할인 쿠폰   
8       2021-03-28 15:26:47               96000   [웰컴쿠폰] 올인원 패키지 2만원 할인 쿠폰   
10      2021-03-28 15:27:17              263000                        NaN   
14      2021-03-28 15:31:01               91000   [웰컴쿠폰] 올인원 패키지 2만원 할인 쿠폰   
...                     ...                 ...                        ...   
105411  2021-09-25 13:39:26              212000    [미래투자금] 온라인 강의 2만원 할인쿠폰   
105412  2021-09-25 13:46:29              149000                        NaN   
105413  2021-09-25 14:01:26              159000    [미래투자금] 온라인 강의 2만원 할인쿠폰   
105416  2021-09-25 14:17:41               96000  [WELCOME] 온라인 강의 2만원 할인쿠폰   
105418  2021-09-25 14:25:47              139000  [WELCOME] 온라인 강의 3만원 할인쿠폰   

        coupon_discount_amount  sale_price  tax_free_amount      pg method  \
2                      19090.0    190900.0         190900.0     NaN   CARD   
4                      20000.0    117500.0         117500.0     NaN   CARD   
8                      20000.0    116000.0         116000.0     NaN   CARD   
10                         NaN    263000.0         263000.0     NaN   CARD   
14                     20000.0    111000.0         111000.0     NaN   CARD   
...                        ...         ...              ...     ...    ...   
105411                 20000.0    232000.0         232000.0  INICIS   CARD   
105412                     NaN    149000.0         149000.0   KAKAO   CARD   
105413                 20000.0    179000.0         179000.0  INICIS   CARD   
105416                 20000.0    116000.0         116000.0  INICIS  VBANK   
105418                 30000.0    169000.0         169000.0   NAVER  POINT   

       subcategory_title  marketing_start_at  
2                    NaN  2021-03-04 0:00:00  
4                PPT/보고서  2019-11-14 0:00:00  
8                  영상 편집  2020-12-02 0:00:00  
10                   NaN  2020-11-27 0:00:00  
14              프론트엔드 개발  2021-02-02 0:00:00  
...                  ...                 ...  
105411              디자인툴  2021-04-23 0:00:00  
105412             영상 편집  2020-12

## 데이터 분포 확인

궁금한 것
- 실제 결제율
- 한사람당 듣는 평균 강의수
- category_title >  subcategory_title > course_title 별 수강생 수 <br>
    - 제일 핫한 카테고리 ? <br>
    - 제일 많이 팔린 강의 ? <br>
- 결제완료자 중 환불 비율<br>
    - 제일 환불 많이 하는 강의?
- 쿠폰의 영향이 있는지 없는지

In [95]:
fc.columns

Index(['id', 'customer_id', 'course_id', 'type', 'state', 'course_title',
       'category_title', 'format', 'completed_at', 'transaction_amount',
       'coupon_title', 'coupon_discount_amount', 'sale_price',
       'tax_free_amount', 'pg', 'method', 'subcategory_title',
       'marketing_start_at'],
      dtype='object')

In [117]:
print(fc['id'].unique())
print(len(fc['id'].unique()))

[ 970657  970658  970669 ... 1364991 1364996 1365007]
62763


In [116]:
print(fc['customer_id'].unique())
print(len(fc['customer_id'].unique()))

[ 72297. 478028. 478033. ... 670891. 640697. 610626.]
42398


In [118]:
print(fc['course_id'].unique())
print(len(fc['course_id'].unique()))

[204246. 201797. 203178. ... 207556. 207731. 207742.]
1311


In [99]:
fc['type'].unique()

array(['PAYMENT', 'REFUND', 'TRANSACTION', 'ORDER'], dtype=object)

In [119]:
print(fc['course_title'].unique())
print(len(fc['course_title'].unique()))

['시크릿코드 : 프론트엔드 실무 완성편' '올인원 패키지 : 직장인 필수 스킬 3종 세트 MAX'
 '초격차 패키지 : 한번에 끝내는 영상 촬영과 편집' ... '프로모션 : 데이터 분석 맥북 air 패스(9월 현물 프리패스)'
 '(B2B) 그린오션스 온라인 구독 올 플랜 - 비대면바우처' '(B2B) (주)오디 온라인 구독 올 플랜 - 비대면바우처']
1592


In [120]:
print(fc['category_title'].unique())
print(len(fc['category_title'].unique()))

['프로그래밍' '업무 생산성' '영상/3D' '부동산/금융' '데이터사이언스' '디자인' '마케팅' '크리에이티브' '투자/재테크'
 '교육' nan '관리' '데이터 사이언스' '비즈니스' '지식콘텐츠' '파이낸스']
16


In [121]:
print(fc['format'].unique())
print(len(fc['format'].unique()))

['스쿨 온라인' '올인원' 'RED' '온라인 완주반' 'B2B 온라인' '캠프' '스쿨' nan 'B2B' '이벤트'
 '하루10분컷' '아카이브' '관리' '온라인' 'MBA' 'B2G']
16


In [122]:
print(fc['completed_at'].unique())
print(len(fc['completed_at'].unique()))

['2021-03-28 15:21:43' '2021-03-28 15:21:54' '2021-03-28 15:26:47' ...
 '2021-09-25 14:01:26' '2021-09-25 14:17:41' '2021-09-25 14:25:47']
60924


In [123]:
print(fc['transaction_amount'].unique())
print(len(fc['transaction_amount'].unique()))

[ 171810   97500   96000 ... -249000 -108186  347650]
2286


In [125]:
print(fc['coupon_title'].unique())
print(len(fc['coupon_title'].unique()))

['[10% 할인] 시크릿코드 실무 완성편' '[웰컴쿠폰] 올인원 패키지 2만원 할인 쿠폰' nan
 '[웰컴쿠폰] 올인원 패키지 1만원 할인 쿠폰' '[웰컴쿠폰] 온라인 완주반 5만원 할인 쿠폰'
 '?즉시할인 쿠폰 코드 : spring5% ?' '[럭키룰렛] 올인원패키지' '타이포그래피 기수강생 할인 쿠폰'
 '[웰컴쿠폰] 올인원 패키지 3만원 할인 쿠폰' '[할인쿠폰] 파이낸스 카테고리 올인원 패키지 30% 할인 (금융 아카데미 제외)'
 '[럭키룰렛] 완주반' '[10% 할인] 프론트엔드 실무 완성편'
 '한 번에 끝내는 Java/Sring 웹 개발 마스터 초격차 패키지 무료 쿠폰 (-2/28)'
 '수강생 50만명 돌파 기념 감사제 쿠폰' '올인원 패키지 : 타이포그래피와 편집디자인 강의 수강권'
 '[무료수강권] 강의 검수용_이호상 PM님' '패캐머_온라인 쿠폰 !'
 '[무료수강권] 파이낸스 카테고리 올인원 패키지_대체투자 자산운용 실무 영상 오픈 일정 변경'
 '한 번에 끝내는 반복 업무 자동화 초격차 패키지 사용 가능 쿠폰' 'crm 검수용 쿠폰 aop'
 '[쿠폰재발급]수강생 50만명 돌파 기념 감사제 쿠폰' '[재발급] 페이백 쿠폰'
 '[무료수강권] 강의검수용_강사님 쿠폰_박혜린 강사님' '신해동 발행 쿠폰' '[무료수강권] 강사님 제공용 RED'
 '[무료수강권] 강사님 제공용' '[할인쿠폰] 영상/디자인 카테고리 올인원 패키지 5만원 할인_영상 오픈 일정 변경'
 '[무료수강권] 내부직원 검수용 RED'
 '[무료수강권] 올인원 패키지 : 인스타를 성장시킨 50명의 성공 방정식 강의검수용_강사님 쿠폰'
 '[무료수강권] 내부직원 강의 검수용' '[무료수강권] 권오상 회계사의 자금조달 분석 및 가치평가_23기 신동현 수강생님'
 '[재발급] [웰컴쿠폰] 올인원 패키지 3만원 할인 쿠폰' '[무료수강권] 완주반 후기 리워드'
 '[할인쿠폰] 초격차 패키지 디자인 툴 5만원 기수강생 할인 쿠폰' 'Vue.js 압축 완성 강의'
 '[무료수강

In [126]:
fc['coupon_discount_amount'].unique()
len(fc['coupon_discount_amount'].unique())

748

In [127]:
fc['sale_price'].unique()
len(fc['sale_price'].unique())

536

In [128]:
fc['tax_free_amount'].unique()
len(fc['tax_free_amount'].unique())

535

In [129]:
print(fc['pg'].unique())
print(len(fc['pg'].unique()))

[nan 'INICIS' 'KAKAO' 'CHAI' 'PAYCO' 'NAVER']
6


In [130]:
print(fc['method'].unique())
print(len(fc['method'].unique()))

['CARD' 'VBANK' 'TRANS' 'PROMOTION' 'TRANSFER' 'ONSITE' 'B2B_CONTRACT'
 'POINT' 'VOUCHER' nan]
10


In [131]:
print(fc['subcategory_title'].unique())
print(len(fc['subcategory_title'].unique()))

[nan 'PPT/보고서' '영상 편집' '프론트엔드 개발' 'DevOps' 'UX/UI' '머신러닝' '모션그래픽'
 '딥러닝/인공지능' '글쓰기/카피라이팅' '디지털마케팅' '엑셀/VBA' '모바일앱' '업무자동화' 'SNS마케팅' '게임'
 '부동산 개발' '데이터분석' '시각디자인' '개발자 커리어' '백엔드 개발' '데이터마케팅' '디자인툴' '금융/투자 실무'
 '회계/재무/세무' '코딩 입문' '커뮤니케이션' '부동산 투자' '리더십' '일러스트' '주식 투자' '게임 그래픽'
 '데이터엔지니어링' 'VFX']
34


In [133]:
print(fc['marketing_start_at'].unique())
print(len(fc['marketing_start_at'].unique()))

['2021-03-04 0:00:00' '2019-11-14 0:00:00' '2020-12-02 0:00:00'
 '2020-11-27 0:00:00' '2021-02-02 0:00:00' nan '2021-02-08 0:00:00'
 '2021-03-11 0:00:00' '2020-12-07 0:00:00' '2021-01-05 0:00:00'
 '2021-01-22 0:00:00' '2020-04-27 0:00:00' '2021-03-19 0:00:00'
 '2021-03-08 0:00:00' '2021-03-10 0:00:00' '2020-12-18 0:00:00'
 '2020-01-04 0:00:00' '2021-01-18 0:00:00' '2020-11-04 0:00:00'
 '2020-02-28 0:00:00' '2021-01-12 0:00:00' '2020-03-31 0:00:00'
 '2021-02-10 0:00:00' '2021-03-16 0:00:00' '2020-01-14 0:00:00'
 '2019-09-30 0:00:00' '2019-05-17 0:00:00' '2021-02-26 0:00:00'
 '2021-02-01 0:00:00' '2019-11-25 0:00:00' '2020-10-23 0:00:00'
 '2020-11-26 0:00:00' '2021-01-28 0:00:00' '2020-05-29 0:00:00'
 '2020-06-19 0:00:00' '2021-02-22 0:00:00' '2020-06-18 0:00:00'
 '2020-04-17 0:00:00' '2019-11-11 0:00:00' '2020-10-06 0:00:00'
 '2020-11-20 0:00:00' '2019-05-02 0:00:00' '2020-09-11 0:00:00'
 '2021-01-04 0:00:00' '2020-04-03 0:00:00' '2021-02-10 18:00:00'
 '2020-05-04 0:00:00' '2021-02-25 0

#### 실결율

In [94]:
# 실결율  
# [1 - {ORDER_CANCELLED / {(ORDER_CANCELLED) + (ORDER_COMPLETED) + (PAYMENT_COMPLETED)}]*100

## ORDER_COMPLETED의 정확한 의미 파악 필요함

print(len(data[(data['type'] == 'ORDER') & (data['state'] == 'CANCELLED')]))
print(len(data[data['type'] == 'ORDER']) + len(data[(data['type'] == 'PAYMENT') & (data['state'] == 'COMPLETED')]))
ap= (1 - (34123/93061))*100
print(round(ap, 2))

34123
93061
63.33


#### 한사람당 수강하는 평균 강의수

In [148]:
fc.groupby("customer_id")
fc.head(8)

id  customer_id  course_id     type      state  \
2   970657      72297.0   204246.0  PAYMENT  COMPLETED   
4   970658     478028.0   201797.0  PAYMENT  COMPLETED   
8   970669     478033.0   203178.0  PAYMENT  COMPLETED   
10  970671     469860.0   203644.0  PAYMENT  COMPLETED   
14  970676     478035.0   203720.0  PAYMENT  COMPLETED   
18  970682     478031.0        NaN  PAYMENT  COMPLETED   
22  970695     455311.0   203146.0  PAYMENT  COMPLETED   
28  970704     471759.0   204068.0  PAYMENT  COMPLETED   

                                      course_title category_title  format  \
2                             시크릿코드 : 프론트엔드 실무 완성편          프로그래밍  스쿨 온라인   
4                    올인원 패키지 : 직장인 필수 스킬 3종 세트 MAX         업무 생산성     올인원   
8                      초격차 패키지 : 한번에 끝내는 영상 촬영과 편집          영상/3D     올인원   
10  올인원 패키지 : 권오상의 금융 아카데미 - 권오상의 재무제표 분석 기초(평생소장)         부동산/금융     올인원   
14                     초격차 패키지 : 한 번에 끝내는 프론트엔드 개발          프로그래밍     올인원   
18           프로모션 : 기초부터 시작하는 데이터분석 한끝세트 (3월 프리패스)        데이터사이언스     올인원   
22                     The RED : 현실 세상의 TDD by 이규원          프로그래밍     RED   
28                      올인원 패키지 : 마디아의 UX/UI 포트폴리오            디자인     올인원   

           completed_at  transaction_amount              coupon_title  \
2   2021-03-28 15:21:43              171810     [10% 할인] 시크릿코드 실무 완성편   
4   2021-03-28 15:21:54               97500  [웰컴쿠폰] 올인원 패키지 2만원 할인 쿠폰   
8   2021-03-28 15:26:47               96000  [웰컴쿠폰] 올인원 패키지 2만원 할인 쿠폰   
10  2021-03-28 15:27:17              263000                       NaN   
14  2021-03-28 15:31:01               91000  [웰컴쿠폰] 올인원 패키지 2만원 할인 쿠폰   
18  2021-03-28 15:33:28              299000                       NaN   
22  2021-03-28 15:43:59              149500                       NaN   
28  2021-03-28 15:46:22               89000  [웰컴쿠폰] 올인원 패키지 1만원 할인 쿠폰   

    coupon_discount_amount  sale_price  tax_free_amount   pg method  \
2                  19090.0    190900.0         190900.0  NaN   CARD   
4                  20000.0    117500.0         117500.0  NaN   CARD   
8                  20000.0    116000.0         116000.0  NaN   CARD   
10                     NaN    263000.0         263000.0  NaN   CARD   
14                 20000.0    111000.0         111000.0  NaN   CARD   
18                     NaN    299000.0         299000.0  NaN   CARD   
22                     NaN    149500.0         149500.0  NaN   CARD   
28                 10000.0     99000.0          99000.0  NaN   CARD   

   subcategory_title  marketing_start_at  
2                NaN  2021-03-04 0:00:00  
4            PPT/보고서  2019-11-14 0:00:00  
8              영상 편집  2020-12-02 0:00:00  
10               NaN  2020-11-27 0:00:00  
14          프론트엔드 개발  2021-02-02 0:00:00  
18               NaN                 NaN  
22            DevOps  2021-02-08 0:00:00  
28             UX/UI  2021-03-11 0:00:00

In [166]:
fc.customer_id.value_counts().mean()

1.480293416987051

In [169]:
fc.customer_id.value_counts()

98685.0     57
427973.0    47
50769.0     42
158374.0    42
3189.0      40
            ..
263854.0     1
441813.0     1
237398.0     1
101851.0     1
609760.0     1
Name: customer_id, Length: 42397, dtype: int64

In [172]:
fc.customer_id.value_counts().loc[fc.customer_id.value_counts() > 1]

98685.0     57
427973.0    47
50769.0     42
158374.0    42
3189.0      40
            ..
576509.0     2
532001.0     2
523630.0     2
518086.0     2
525142.0     2
Name: customer_id, Length: 11500, dtype: int64

In [180]:
print(11500/42397*100)

27.124560700049532


- 일인당 1.48개 강의 수강 (약 2개 강의 평균으로 보임) <br>
- 2개 강의 이상 수강하는 수강생은 결제한 수강생 중 약 27%


In [209]:
#customers = fc.customer_id.value_counts().loc[fc.customer_id.value_counts() > 1].index
#fc.loc[fc.customer_id.isin(customers)]

id  customer_id  course_id     type      state  \
2        970657      72297.0   204246.0  PAYMENT  COMPLETED   
14       970676     478035.0   203720.0  PAYMENT  COMPLETED   
28       970704     471759.0   204068.0  PAYMENT  COMPLETED   
40       970730     445723.0   203720.0  PAYMENT  COMPLETED   
46       970740     474451.0   202883.0  PAYMENT  COMPLETED   
...         ...          ...        ...      ...        ...   
105393  1364869     638903.0   203178.0  PAYMENT  COMPLETED   
105398  1364886     578540.0   202789.0  PAYMENT  COMPLETED   
105399  1364889      46138.0   206224.0  PAYMENT  COMPLETED   
105406  1364929     120686.0        NaN  PAYMENT  COMPLETED   
105412  1364976     384003.0   203178.0  PAYMENT  COMPLETED   

                            course_title category_title  format  \
2                   시크릿코드 : 프론트엔드 실무 완성편          프로그래밍  스쿨 온라인   
14           초격차 패키지 : 한 번에 끝내는 프론트엔드 개발          프로그래밍     올인원   
28            올인원 패키지 : 마디아의 UX/UI 포트폴리오            디자인     올인원   
40           초격차 패키지 : 한 번에 끝내는 프론트엔드 개발          프로그래밍     올인원   
46            올인원 패키지 : 디지털 마케팅 MAX 2020            마케팅     올인원   
...                                  ...            ...     ...   
105393       초격차 패키지 : 한번에 끝내는 영상 촬영과 편집          영상/3D     올인원   
105398  올인원 패키지 : 2021 취업/이직/투자를 위한 기업분석         투자/재테크     올인원   
105399      초격차 패키지 : 한 번에 끝내는 부동산 경매/공매         투자/재테크     올인원   
105406        [한정판매] 자연어 A-Z 끝장 패키지(~9월)        데이터사이언스     올인원   
105412       초격차 패키지 : 한번에 끝내는 영상 촬영과 편집          영상/3D     올인원   

               completed_at  transaction_amount               coupon_title  \
2       2021-03-28 15:21:43              171810      [10% 할인] 시크릿코드 실무 완성편   
14      2021-03-28 15:31:01               91000   [웰컴쿠폰] 올인원 패키지 2만원 할인 쿠폰   
28      2021-03-28 15:46:22               89000   [웰컴쿠폰] 올인원 패키지 1만원 할인 쿠폰   
40      2021-03-28 16:15:46               91000   [웰컴쿠폰] 올인원 패키지 2만원 할인 쿠폰   
46      2021-03-28 16:28:42              114500   [웰컴쿠폰] 올인원 패키지 2만원 할인 쿠폰   
...                     ...                 ...                        ...   
105393  2021-09-25 11:22:07              139000  [WELCOME] 온라인 강의 1만원 할인쿠폰   
105398  2021-09-25 11:54:34              145000                        NaN   
105399  2021-09-25 11:55:19              189000                        NaN   
105406  2021-09-25 12:29:39              481000                        NaN   
105412  2021-09-25 13:46:29              149000                        NaN   

        coupon_discount_amount  sale_price  tax_free_amount      pg method  \
2                      19090.0    190900.0         190900.0     NaN   CARD   
14                     20000.0    111000.0         111000.0     NaN   CARD   
28                     10000.0     99000.0          99000.0     NaN   CARD   
40                     20000.0    111000.0         111000.0     NaN   CARD   
46                     20000.0    134500.0         134500.0     NaN   CARD   
...                        ...         ...              ...     ...    ...   
105393                 10000.0    149000.0         149000.0   KAKAO  POINT   
105398                     NaN    145000.0         145000.0   KAKAO  POINT   
105399                     NaN    189000.0         189000.0  INICIS   CARD   
105406                     NaN    481000.0         481000.0  INICIS   CARD   
105412                     NaN    149000.0         149000.0   KAKAO   CARD   

       subcategory_title  marketing_start_at  
2                    NaN  2021-03-04 0:00:00  
14              프론트엔드 개발  2021-02-02 0:00:00  
28                 UX/UI  2021-03-11 0:00:00  
40              프론트엔드 개발  2021-02-02 0:00:00  
46                디지털마케팅                 NaN  
...                  ...                 ...  
105393             영상 편집  2020-12-02 0:00:00  
105398             주식 투자  2020-09-11 0:00:00  
105399            부동산 투자  2021-08-23 0:00:00  
105406               NaN                 NaN  
105412             영상 편집  2020-12-02 0:00:00  

[31863 rows 

#### category_title > subcategory_title > course_title 별 수강생 수
제일 핫한 카테고리 ?
제일 많이 팔린 강의 ?

In [189]:
categories = fc[['category_title', 'subcategory_title', 'course_title', 'customer_id']]
categories

category_title subcategory_title  \
2               프로그래밍               NaN   
4              업무 생산성           PPT/보고서   
8               영상/3D             영상 편집   
10             부동산/금융               NaN   
14              프로그래밍          프론트엔드 개발   
...               ...               ...   
105411            디자인              디자인툴   
105412          영상/3D             영상 편집   
105413          프로그래밍            백엔드 개발   
105416          영상/3D             영상 편집   
105418          영상/3D             영상 편집   

                                          course_title  customer_id  
2                                 시크릿코드 : 프론트엔드 실무 완성편      72297.0  
4                        올인원 패키지 : 직장인 필수 스킬 3종 세트 MAX     478028.0  
8                          초격차 패키지 : 한번에 끝내는 영상 촬영과 편집     478033.0  
10      올인원 패키지 : 권오상의 금융 아카데미 - 권오상의 재무제표 분석 기초(평생소장)     469860.0  
14                         초격차 패키지 : 한 번에 끝내는 프론트엔드 개발     478035.0  
...                                                ...          ...  
105411                      초격차 패키지 : 한번에 끝내는 3D 디자인 툴     669238.0  
105412                     초격차 패키지 : 한번에 끝내는 영상 촬영과 편집     384003.0  
105413                     초격차 패키지 : 한 번에 끝내는 파이썬 웹 개발     670891.0  
105416                 초격차 패키지 : 한 번에 끝내는 유튜브 크리에이터 되기     640697.0  
105418                              용호수의 돈 버는 실전 영상 제작     610626.0  

[62763 rows x 4 columns]

In [214]:
fc.category_title.value_counts()

프로그래밍       20957
데이터사이언스      8193
업무 생산성       7500
디자인          6942
영상/3D        6100
마케팅          5956
부동산/금융       3312
투자/재테크       1735
크리에이티브       1044
교육            934
지식콘텐츠          33
비즈니스           24
파이낸스           19
데이터 사이언스        5
관리              1
Name: category_title, dtype: int64

In [219]:
a = fc.groupby('category_title')
a.head()

id  customer_id  course_id     type      state  \
2       970657      72297.0   204246.0  PAYMENT  COMPLETED   
4       970658     478028.0   201797.0  PAYMENT  COMPLETED   
8       970669     478033.0   203178.0  PAYMENT  COMPLETED   
10      970671     469860.0   203644.0  PAYMENT  COMPLETED   
14      970676     478035.0   203720.0  PAYMENT  COMPLETED   
...        ...          ...        ...      ...        ...   
80763  1247605     612513.0   206638.0  PAYMENT  COMPLETED   
84030  1264094     202885.0        NaN  PAYMENT  COMPLETED   
84250  1265593     621957.0        NaN  PAYMENT  COMPLETED   
84788  1269086     624258.0        NaN  PAYMENT  COMPLETED   
84871  1269667     614783.0        NaN  PAYMENT  COMPLETED   

                                         course_title category_title   format  \
2                                시크릿코드 : 프론트엔드 실무 완성편          프로그래밍   스쿨 온라인   
4                       올인원 패키지 : 직장인 필수 스킬 3종 세트 MAX         업무 생산성      올인원   
8                         초격차 패키지 : 한번에 끝내는 영상 촬영과 편집          영상/3D      올인원   
10     올인원 패키지 : 권오상의 금융 아카데미 - 권오상의 재무제표 분석 기초(평생소장)         부동산/금융      올인원   
14                        초격차 패키지 : 한 번에 끝내는 프론트엔드 개발          프로그래밍      올인원   
...                                               ...            ...      ...   
80763         (B2B) 주식회사 에쓰엠에쓰 온라인 구독 패키지 플랜 - 비대면바우처           파이낸스  B2B 온라인   
84030     프로모션 : 시장의 흐름을 읽는 투자의 정석 평생 PASS  (8월 프리패스)           파이낸스      올인원   
84250     프로모션 : 시장의 흐름을 읽는 투자의 정석 평생 PASS  (8월 프리패스)           파이낸스      올인원   
84788     프로모션 : 시장의 흐름을 읽는 투자의 정석 평생 PASS  (8월 프리패스)           파이낸스      올인원   
84871     프로모션 : 시장의 흐름을 읽는 투자의 정석 평생 PASS  (8월 프리패스)           파이낸스      올인원   

              completed_at  transaction_amount              coupon_title  \
2      2021-03-28 15:21:43              171810     [10% 할인] 시크릿코드 실무 완성편   
4      2021-03-28 15:21:54               97500  [웰컴쿠폰] 올인원 패키지 2만원 할인 쿠폰   
8      2021-03-28 15:26:47               96000  [웰컴쿠폰] 올인원 패키지 2만원 할인 쿠폰   
10     2021-03-28 15:27:17              263000                       NaN   
14     2021-03-28 15:31:01               91000  [웰컴쿠폰] 올인원 패키지 2만원 할인 쿠폰   
...                    ...                 ...                       ...   
80763  2021-07-28 15:04:48              480000                       NaN   
84030   2021-08-04 7:13:32              330000                       NaN   
84250  2021-08-04 22:27:15              330000                       NaN   
84788  2021-08-06 17:47:15              330000                       NaN   
84871   2021-08-07 1:32:51              330000                       NaN   

       coupon_discount_amount  sale_price  tax_free_amount      pg    method  \
2                     19090.0    190900.0         190900.0     NaN      CARD   
4                     20000.0    117500.0         117500.0     NaN      CARD   
8                     20000.0    116000.0         116000.0     NaN      CARD   
10                        NaN    263000.0         263000.0     NaN      CARD   
14                    20000.0    111000.0         111000.0     NaN      CARD   
...                       ...         ...              ...     ...       ...   
80763                     NaN         NaN              NaN     NaN  TRANSFER   
84030                     NaN    330000.0         330000.0   NAVER      CARD   
84250                     NaN    330000.0         330000.0  INICIS      CARD   
84788                     NaN    330000.0         330000.0   NAVER      CARD   
84871                     NaN    330000.0         330000.0   NAVER     POINT   

      subcategory_title  marketing_start_at  
2                   NaN  2021-03-04 0:00:00  
4               PPT/보고서  2019-11-14 0:00:00  
8                 영상 편집  2020-12-02 0:00:00  
10                  NaN  2020-11-27 0:00:00  
14             프론트엔드 개발  2021-02-02 0:00:00  
...                 ...                 ...  
80763               NaN                 NaN  
84030               NaN                 NaN  
84250 

#### 결제완료자 중 환불 비율
제일 환불 많이 하는 강의?

#### 쿠폰의 영향이 있는지 없는지